HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,354 @@
+  }
+
+  if (maxDeltaR_L2L3_ <= 0.) {
+    throw cms::Exception("Configuration")
+        << "invalid value for parameter \"MaxDeltaRL2L3\" (must be > 0): " << maxDeltaR_L2L3_;
+  }
+}
+
+void HLTMuonL3orL2PreFilter::MuonSelectionCuts::fillPSetDescription(edm::ParameterSetDescription& desc) {
+  desc.add<int>("MinN", 1);
+  desc.add<double>("MaxDXYBeamSpot", 9999.0);
+  desc.add<double>("MinDXYBeamSpot", -1.0);
+  desc.add<double>("MaxDz", 9999.0);
+  desc.add<double>("MinDxySig", -1.0);
+  desc.add<double>("NSigmaPt", 0.0);
+  desc.add<double>("MaxPtDifference", 9999.0);
+  desc.add<double>("MinTrackPt", 0.0);
+  desc.add<double>("MaxNormalizedChi2", 9999.0);
+  desc.add<double>("MinPt", 23.0);
+  desc.add<double>("MaxEta", 2.5);
+  desc.add<int>("MinNmuonHits", 0);
+  desc.add<std::vector<int>>("MinNhits", {1, 0});
+  desc.add<std::vector<double>>("AbsEtaBins", {1, 9999.});
+  desc.add<std::vector<int>>("MinNstations", {1, 1});
+  desc.add<std::vector<int>>("MinNchambers", {1, 0});
+};
+
+void HLTMuonL3orL2PreFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  makeHLTFilterDescription(desc);
+  desc.add<edm::InputTag>("L3CandTag", edm::InputTag("hltL3MuonCandidates"));
+  desc.add<edm::InputTag>("L2CandTag", edm::InputTag("hltL2MuonCandidates"));
+  desc.add<edm::InputTag>("PreviousCandTag", edm::InputTag(""));
+  desc.add<edm::InputTag>("BeamSpotTag", edm::InputTag("hltOnlineBeamSpot"));
+  desc.add<bool>("MatchToPreviousCand", true);
+  desc.add<double>("MaxDeltaRL2L3", 0.05);
+
+  edm::ParameterSetDescription descMuonSelL2;
+  MuonSelectionCuts::fillPSetDescription(descMuonSelL2);
+  desc.add<edm::ParameterSetDescription>("L2CandSelection", descMuonSelL2);
+
+  edm::ParameterSetDescription descMuonSelL3;
+  MuonSelectionCuts::fillPSetDescription(descMuonSelL3);
+  desc.add<edm::ParameterSetDescription>("L3CandSelection", descMuonSelL3);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+  if (saveTags()) {
+    filterproduct.addCollectionTag(L3CandTag_);
+    filterproduct.addCollectionTag(L2CandTag_);
+  }
+  // get hold of trks
+  auto const L3mucands = iEvent.getHandle(L3CandToken_);
+  auto const L2mucands = iEvent.getHandle(L2CandToken_);
+  auto const previousLevelCands = iEvent.getHandle(previousCandToken_);
+  std::vector<reco::RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(trigger::TriggerMuon, vcands);
+  }
```

Comment:
```suggestion
  std::vector<reco::RecoChargedCandidateRef> vcands;
  if (matchPreviousCand_) {
    auto const& previousLevelCands = iEvent.get(previousCandToken_);
    previousLevelCands.getObjects(trigger::TriggerMuon, vcands);
  }
```
I noticed that the vector needs to be filled only if `matchPreviousCand_ == true`. And when that is the case, the collection should be always valid; if not, the fwk should return an error (this is what the suggested update does, the validity check is done inside `iEvent.get`).